caom2pipe/tests/test_reporting.py

```diff
@@ -0,0 +1,330 @@
+from test_run_composable import TEST_BOOKMARK, _write_state
+
+
+def test_report_output_todo_local(test_config):
+    test_config.cleanup_files_when_storing = True
+    test_config.store_modified_files_only = True
+    # work is all the list of files,
+    #   one has been stored already,
+    #   one fails fitsverify
+    #   one fails ingestion
+    #   one succeeds
+
+    test_config.logging_level = 'INFO'
+    test_config.task_types = [TaskType.STORE]
+    test_config.interval = 100
+
+    pre_success_listing, file_info_list = _create_dir_listing('/tmp', 4)
+    test_reader = Mock()
+
+    test_start_time = datetime(year=2020, month=3, day=3, hour=1, minute=1, second=1)
+    test_end_time = datetime(year=2020, month=3, day=3, hour=2, minute=1, second=1)
+
+    store_modified_true_move_calls = [
+        call('/tmp/A0.fits', '/data/success'),
+        call('/tmp/A1.fits', '/data/failure'),
+        call('/tmp/A2.fits', '/data/failure'),
+        call('/tmp/A3.fits', '/data/success'),
+    ]
+
+    store_modified_false_move_calls = [
+        call('/tmp/A1.fits', '/data/failure'),
+        call('/tmp/A0.fits', '/data/success'),
+        call('/tmp/A2.fits', '/data/failure'),
+        call('/tmp/A3.fits', '/data/success'),
+    ]
+
+    store_modified_true_put_side_effect = [CadcException, None]
+    store_modified_false_put_side_effect = [None, CadcException, None]
+
+    clean_up_false_put_side_effect = [None, CadcException, None]
+
+    store_modified_true_info_side_effect = [file_info_list[0], None, None, None, file_info_list[0]]
+    store_modified_false_info_side_effect = [
+        file_info_list[0], None, file_info_list[0], file_info_list[0], file_info_list[0]
+    ]
+
+    store_modified_true_file_info_get_side_effect = [file_info_list[0], file_info_list[3], None, None]
+    store_modified_false_file_info_get_side_effect = [file_info_list[0], file_info_list[2], file_info_list[3]]
+
+    store_modified_true_skipped_sum = 1
+    store_modified_false_skipped_sum = 0
+
+    vo_listing, vo_file_info_list, vo_listdir = _get_node_listing(4)
+    vo_client_listing_side_effect = vo_listdir
+    vo_client_get_node_side_effect = vo_listing
+
+    test_clients = Mock()
+    with TemporaryDirectory() as temp_dir:
+        test_config.working_directory = temp_dir
+        test_config.log_file_directory = temp_dir
+        test_config.state_fqn = f'{temp_dir}/{test_config.state_file_name}'
+        test_config.log_file_directory = temp_dir
+        test_config.failure_log_file_name = 'fail.txt'
+        test_config.retry_file_name = 'do_again.txt'
+        test_config.success_log_file_name = 'good.txt'
+        test_config.progress_file_name = 'progress.txt'
+        test_config.rejected_file_name = 'reject.txt'
+        test_config._report_fqn = f'{temp_dir}/report.txt'
+
+        for clean_up in [True, False]:
+            for store_modified in [True, False]:
+                move_calls = store_modified_false_move_calls
```

Comment:
It's a bit hard to follow the logic with these loops that use variables initiated earlier.

Reply:
The other choice is to write multiple tests with lots of only-mostly duplicate code. Is that a better approach? I could be missing something when it comes to function-level fixtures as a solution for this kind of integration test.

Reply:
Reorganized this as much as it seem non-repetitive to do. Let me know if it's better.